### DISCUSSION TITLE
Community Feedback Requested: WinUI Controls for Avalonia.

### DISCUSSION THREAD

**Discussion by MikeCodesDotNET** [AvaloniaUI/Avalonia#5893] [Ideas]
Hey! 

I've recently seen the incredible work of @amwx, who has been [porting WinUI controls](https://github.com/amwx/FluentAvalonia) over to Avalonia. This has started a discussion about which (if any) WinUI controls should be ported and included in Avalonia and how should we prioritise them. 

I've included below some WinUI controls that would be helpful to have in Avalonia to get things started. Feel free to add any I've missed (only WinUI controls), and you want to see ported. 

To cast your vote, please thumbs up (👍) or thumbs down (👎). 

Your feedback is incredibly valuable for us to understand the community's desires, so please vote and share your thoughts!

**Comment by MikeCodesDotNET** [comment]
## ColorPicker

![](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls/images/controls/color-picker-default.png?view=winrt-19041)

[ColorPicker API](https://docs.microsoft.com/en-gb/uwp/api/Windows.UI.Xaml.Controls.ColorPicker?view=winrt-19041)

**Reply by robloo** [reply]
See #4179.

The plan is to use @maxkatz6 existing port but not to follow WinUI's (in my opinion) ugly style. Instead, use the latest ColorPicker and ColorPickerButton styles from the UWP Community Toolkit: https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3363. This also supports color palette's and sliders for each color channel. ~It will likely be 100% API compatible with the WinUI control -- just with a new Style, new functionality and a new Button variant.~

**Reply by robloo** [reply]
I should add that I signed up for this one too. If there are individuals that actually need it now that would help prioritize. Getting it widely tested is important and if there are people to test in their apps I could likely port it over soon.

I will need it for a commercial app in the future but that is likely several months if not a year away. Getting it stabilized by then would be helpful though.

**Reply by Splitwirez** [reply]
Sorry, what exactly is the UWP Community Toolkit variant supposed to look like? I'm guessing this is not what you had in mind:
![image](https://user-images.githubusercontent.com/34009058/120336161-1498ce00-c2c0-11eb-93b5-b645e8b331c4.png)

**Reply by robloo** [reply]
@Splitwirez Thats a browser compatibility issue, just use a different one. Graphics are in the description at the very top.

**Reply by robloo** [reply]
![image](https://user-images.githubusercontent.com/17993847/120359846-24231180-c2d6-11eb-9bb7-2701733db9d7.png)

**Reply by timunie** [reply]
I also need a color picker. If there is something to test, let me know 😉

I would be happy to help

**Reply by jp2masa** [reply]
You can see the current progress on this issue: https://github.com/AvaloniaUI/Avalonia/issues/4179#issuecomment-1054653895.

**Reply by robloo** [reply]
Completed with #8215, feel free to test

**Comment by MikeCodesDotNET** [comment]
## RatingControl 

![image](https://user-images.githubusercontent.com/552074/117642051-cd1a9880-b17e-11eb-8efc-280469aafde9.png)

[RatingControl API](https://docs.microsoft.com/en-gb/uwp/api/Windows.UI.Xaml.Controls.RatingControl?view=winrt-19041)

**Reply by maxkatz6** [reply]
This control is very limited in WinUI and wasn't updated in ages.

**Reply by Splitwirez** [reply]
Might bringing `RatingControl` to Avalonia make for a good April Fools' Day joke? /s

**Reply by DmitriyYukhanov** [reply]
This nice `TemplatedControl` [tutorial](https://github.com/timunie/Avalonia.Samples/tree/Controls/RatingControl/src/Avalonia.Samples/CustomControls/RatingControlSample) by @timunie does the job, actually! 

Needs just a fraction of changes to work with Avalonia 11 Preview 4 and don't forget to [include style](https://github.com/AvaloniaUI/Avalonia/issues/6475) into the `App.axaml`, I've lost an hour trying to figure out why control is not visible in the app while being visible in the Preview =D

Here is how you could extend and customize it, for example:
![ratecontrol](https://user-images.githubusercontent.com/883587/217318672-76db27ec-ebd0-427d-bf82-f56ace4efd79.gif)

**Comment by MikeCodesDotNET** [comment]
## SplitButton 

![image](https://user-images.githubusercontent.com/552074/117643324-2f27cd80-b180-11eb-83fc-30bc2d25929f.png)

[SplitButton API](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls.splitbutton?view=winrt-19041)

**Reply by Splitwirez** [reply]
I saw an issue post a while back from someone who'd wanted to look into this, along with its toggle-able counterpart, but was waiting on Flyout...now that `Flyout` is taken care of, maybe it's time to go let them know. I'll go see to that, and let them know to follow up here if they're still able to do the thing.

**Reply by robloo** [reply]
Yes, I signed up for this one too. It should be quite easy considering all the code floating around. I will try to reach out to @amwx and work out a good strategy.

**Reply by robloo** [reply]
Update: This was added to master in PR #7422

**Comment by MikeCodesDotNET** [comment]
## Flyout 

![](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls/images/controls/flyout.png?view=winrt-19041)

[Flyout API](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls.flyout?view=winrt-19041)

**Reply by maxkatz6** [reply]
Already done

**Comment by MikeCodesDotNET** [comment]
## PersonPicture

![image](https://user-images.githubusercontent.com/552074/117642682-79f51580-b17f-11eb-9063-a782ee17fce8.png)

[PersonPicture API](https://docs.microsoft.com/en-gb/uwp/api/Windows.UI.Xaml.Controls.PersonPicture?view=winrt-19041)

**Reply by robloo** [reply]
I'll actually downvote this one as I think it's far too specialized and application specific. A control for displaying a user identify, really? Even if this was an attempt by Microsoft to standardize on how user identities are pictured on Windows it certainly doesn't apply cross-platform.

**Reply by timunie** [reply]
I think this one is easy to implement via DataTemplate.

**Comment by MikeCodesDotNET** [comment]
## NavigationView 

![image](https://user-images.githubusercontent.com/552074/117642779-9bee9800-b17f-11eb-98b0-757d799b1bd4.png)

[NavigationView API](https://docs.microsoft.com/en-gb/uwp/api/Windows.UI.Xaml.Controls.NavigationView?view=winrt-19041)

**Reply by robloo** [reply]
See #2521 

There are multiple avenues for this: 

 1. ModerWpf has an early port directly from WinUI for WPF (maybe the first one). This pre-dates the hierarchical menus I believe and several of the other more recent changes changes (I'm not sure if @Kinnara had a chance to make these updates with depend on SelectionModel and ItemsRepeater).
 2. The Uno Platform also completed a brand-new pure C++/WinRT to C#/WinRT port some months ago. This was the latest C# port I was aware of but @amwx now likely supersedes this. This control can only be used for reference as it is not license-compatible with Avalonia.
 3. @amwx Likely now has the definitive port that should be used for this when it's ready. We can refer to ports one and two above if there are any questions/concerns.

I am curious if all the advanced hierarchical menus are included with port three (@amwx) as that requires 'tree' support in SelectionModel with was not implemented in Avalonia. Plan was a separate TreeSelectionModel or similar or just implement it for the control only.

As so many people/companies have asked for this control I think it would be accepted (#2521 states as much). However, it is quite large and highly specialized so someone would likely need to agree to maintain it.

**Reply by amwx** [reply]
@robloo My port is basically up to date with the current version in WinUI and is a direct C++ to C#/Avalonia port (as best as that can be). I took the original SelectionModel port and used that, so yes hierarchical and all is already there. The only caveat is that my lib defaults to the new WinUI styles (although, I have made it so it can still be used with the base Fluent theme in Avalonia without needing to adopt the new styles). Also as pointed out in gitter last week, multiple data templates (DataTemplateSelector equivalent) aren't supported yet.

**Comment by MikeCodesDotNET** [comment]
## ProgressRing 

![image](https://user-images.githubusercontent.com/552074/117643017-de17d980-b17f-11eb-83f0-11bcb761a255.png)

[ProgressRing API](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls.progressring?view=winrt-19041)

**Reply by robloo** [reply]
This one is tricky. The new WinUI ProgressRing is using Lottie animations. I see Avalonia has a Lottie project but it is not license compatible with the main Avalonia UI repo.

AnimatedIcon will probably end up in the same position in the future.

**Reply by Splitwirez** [reply]
brb taking a crack at doing this without Lottie

**Reply by maxkatz6** [reply]
Lottie in both UWP and Avalonia implementations has better theoretical performance because it's executed on render thread and does not work with xaml logical tree, unlikely simple animations in both frameworks.

On other hand Lottie controls are less customizable.

**Reply by Splitwirez** [reply]
PR for this one incoming, no Lottie lib required.

**Reply by Splitwirez** [reply]
Progress bagels, fresh out of the oven: https://github.com/AvaloniaUI/Avalonia/pull/6003

**Reply by robloo** [reply]
This control now seems in limbo. Past PR's were closed and there is no response in #7716

**Comment by MikeCodesDotNET** [comment]
## TeachingTip 

![image](https://user-images.githubusercontent.com/552074/117643196-08699700-b180-11eb-9139-ea33cddd8b99.png)

[TeachingTip API](https://docs.microsoft.com/en-gb/windows/winui/api/microsoft.ui.xaml.controls.teachingtip?view=winui-3.0)

**Reply by pr8x** [reply]
Seems like some flyout with extra steps. Not sure if this is general-purpose enough.

**Reply by robloo** [reply]
Well, I think its an attempt to standardize this concept more for end users. Having tips displayed in a standard way makes things more user friendly and familiar which is especially important in the case of teaching. Otherwise, I agree this may be too specialized.

**Comment by MikeCodesDotNET** [comment]
## ToggleSplitButton 

![image](https://user-images.githubusercontent.com/552074/117643378-3d75e980-b180-11eb-914e-4ebd69fc621d.png)

[ToggleSplitButton API](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls.togglesplitbutton?view=winrt-19041)

**Reply by robloo** [reply]
See #4959 and #4960

This would come over the same time as SplitButton and likely DropDownButton. All buttons are fair game in my opinion and it's on my todo list.

**Reply by robloo** [reply]
Update: This was added to master in PR https://github.com/AvaloniaUI/Avalonia/pull/7422

**Comment by MikeCodesDotNET** [comment]
## CommandBar 

![image](https://user-images.githubusercontent.com/552074/117643677-8cbc1a00-b180-11eb-90eb-d776ab1516cb.png)

[CommandBar API](https://docs.microsoft.com/en-gb/uwp/api/windows.ui.xaml.controls.commandbar?view=winrt-19041)

**Reply by robloo** [reply]
No public code for this exists yet for WinUI. Everyone has separate implementations (if they are implemented at all). I expect when WinUI3 open sources that will quickly change. It might be best to wait a little while on this one.

**Reply by robloo** [reply]
Another question with this: CommandBar contains: `AppBarButton`, `AppBarToggleButton` and `AppBarSeparator`. The naming of these controls is a relic from the Win8 days and the deprecated app bar. It probably isn't a good idea to carry these names over to Avalonia UI. It would also be great if the control was generalized more and took existing buttons, toggle buttons, etc. Avalonia has other menu types that perhaps could be re-used for this (or generalized to support this case).

**Reply by Splitwirez** [reply]
I *partially* agree that staying in line with WinUI 3's take would be desirable...I say "partially" because I can't help but feel that this argument would likely be a moot point if Avalonia were to rename the `AppBar*` controls to something more sensible like `CommandBar*` as @robloo mentioned, or even more so if Avalonia were to entirely forgo having "special" button controls for the `CommandBar` at all. After all, these would already be incompatible changes, so...just how much would being *less* or *more* incompatible even matter at that point, really? (That said, I am open to being proven wrong)

Also, with the help of Avalonia's style selectors, using the regular button controls would be fairly easy, no?


And another thing...I already have a `CommandBar` implementation in my  [Mechanism](https://github.com/Splitwirez/Mechanism-for-Avalonia) controls library (albeit one modelled more closely after the CommandBar seen in Windows Vista and 7's File Explorer than its WinUI contemporary)...
![Mechanism CommandBar in action](https://raw.githubusercontent.com/Splitwirez/Mechanism-for-Avalonia/master/ReadmeImages/CommandBar.gif)
Barring my potentially-unwarranted code quality concerns, I would be honoured to donate my implemenation to Avalonia itself, tweaked as necessary. (For context, the "layers" feature is implemented entirely separately from the "move the things into the flyout if there isn't enough space to show them in the main body of the control itself" part, so if layers are undesirable for inclusion in Avalonia, that would be very easy to omit.)

**Reply by robloo** [reply]
The AppBar* controls for command bar have a special Icon and Label property. These properties make a lot of sense in the context of menus. Loosing them means everyone has to do it themselves. So we would trade speed of development for generalization.

**Reply by Splitwirez** [reply]
@robloo That's a fair point...but perhaps I should've mentioned that my `CommandBar` uses an attached property for button icons, to achieve the best of both worlds. That same attached property is also used in a number of other places.

**Reply by robloo** [reply]
This exists in FluentAvalonia now.

**Comment by robloo** [comment]
## ContentDialog

![image](https://user-images.githubusercontent.com/17993847/120234147-5e80a600-c225-11eb-9df2-3e9050daaafc.png)

[ContentDialog API](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.contentdialog?view=winrt-20348)

**Reply by maxkatz6** [reply]
Can be considered as part of this feature, which also requires managed fallback dialog implementation
https://github.com/AvaloniaUI/Avalonia/issues/670

**Reply by robloo** [reply]
I would argue this should ONLY be implemented as a managed overlay: the same as UWP. This makes the most sense for all platforms to be easily supported including mobile. A NativeDialog would be another control IMO.

**Reply by Splitwirez** [reply]
@robloo I'm in agreement with that as valid and valuable option, though I do think that "modal popup window where possible, managed overlay otherwise" is a valid use case as well, so perhaps it might be worthwhile to offer both behaviours somehow? Maybe even as another `AppBuilder` extension, similar to enabling the use of managed file dialogs?

Though that may still leave the question of which should be the default behaviour...

**Comment by robloo** [comment]
## AnimatedIcon

![](https://user-images.githubusercontent.com/47644625/86175863-0a8baf80-baf2-11ea-836c-753d868cb01a.gif)

https://github.com/microsoft/microsoft-ui-xaml/issues/2802

[Pre-release AnimatedIcon API](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.animatedicon?view=winui-2.6-prerelease)

**Reply by maxkatz6** [reply]
As part of build-in icons feature https://github.com/AvaloniaUI/Avalonia/issues/4087
Also requires  https://github.com/AvaloniaUI/Avalonia.Lottie

**Reply by Splitwirez** [reply]
...would these Lottie animations use resources (brushes, colours, etc) defined within an Avalonia application? I've been using XAML animations on what may well be the weirdest `TemplatedControl`s known to humankind to do animated icons, to achieve results similar to ![](https://i.imgur.com/7T04Nd9.png) (admittedly this example is in WPF, but Avalonia can pull it off just as well, if not better, than WPF). I, for one, would be hesitant to use Lottie animations for animated icons and such if Avalonia app-defined resources would be off the table as a result.

Just food for thought, I suppose...I doubt many other devs are insane enough to do it this way, lol.

**Reply by robloo** [reply]
I raised a lot of similar concerns with Lottie in the original discussion. I would prefer XAML itself is updated to support all these things. If not updating StoryBoard XAML Animations, the integration with Lottie needs to be much further improved and generalized. However, I quickly lost that discussion and everyone is switching over to Lottie for even determinate progress scenarios (as I predicted at the time).

https://github.com/microsoft/microsoft-ui-xaml/pull/1858#issuecomment-583792001

and others

https://github.com/microsoft/microsoft-ui-xaml/pull/1858#issuecomment-583026692
https://github.com/microsoft/microsoft-ui-xaml/pull/2584#issuecomment-648338408
https://github.com/microsoft/microsoft-ui-xaml/issues/1170

**Reply by jmacato** [reply]
I must say.. with my experience of porting Lottie over to Avalonia.. I just think it's insane that they are using it for even the most banal stuff. It's great for large and complex animations but a progressbar? srsly?

**Reply by Splitwirez** [reply]
Yeah, just because they're taking what they see as the path of least resistance doesn't mean it's a good idea for Avalonia to do the same. Heck, the entire premise of Avalonia itself kind of flies in the face of the notion of "taking the path of least resistance", no?

**Reply by robloo** [reply]
Microsoft is also getting into the "meta" programming/design concepts. They are using a single design source to generate styles and design elements for many frameworks at once. Because Lottie is well supported on the web and mobile is probably why they went that direction for WinUI too. I certainly agree if we can avoid it we should though.

**Reply by Splitwirez** [reply]
> I certainly agree if we can avoid it we should though.

@robloo Then I guess it's a good thing I've already set out to do the `ProgressRing` for Avalonia without relying on Lottie, huh?

**Comment by robloo** [comment]
## MapControl

I'll add this one with a big caveat: I don't expect Avalonia to roll it's own MapControl. However, there should be a partnership with Mapsui in my opinion to get a MapControl ready-to-use with Avalonia.

**Reply by maxkatz6** [reply]
There is a port of Mapsui for Avalonia, but it wasn't completed AFAIK https://github.com/Mapsui/Mapsui/issues/749

**Reply by maxkatz6** [reply]
Also using native map controls for Win and macOS shouldn't be a huge problem since we already have NativeControlHost implemented (example https://github.com/AvaloniaUI/Avalonia/blob/master/samples/interop/NativeEmbedSample/EmbedSample.cs)

But there is another problem with Windows API itself. It's not easily possible to use UWP control implementations (neither XAML Islands) in .NET 5 because it was completely broken with new CsWinRT. And for CsWinRT there is no any ready to use maps implementation either. I don't remember if there were older WinForms/WPF map controls though. Some third parties might work just fine.

**Reply by robloo** [reply]
There are still several blocking missing features to try out WinUI3 and CsWinRT on my end. I know there are a lot of pain points though - and unfixable performance degradations - but I don't have first hand experience with it yet. If you are telling me that they don't have WinUI3 + .NET 5/6 working well that's not good news. I was expecting to transition over to a desktop app model and drop UWP because I need the latest versions of C#/.NET... If I could skip it altogether and go for Avalonia I'll do it, but Avalonia just isn't quite there yet either. 

As far as the MapControl I already opened https://github.com/Mapsui/Mapsui/issues/1125 and plan to support that. I am aware an Avalonia variation of the control was started: would just be great to get an official collaboration. That said, I was not aware you have native control embedding working to the point it's ready for production. If that is ready-to-go for embedding macOS Webviews and mapcontrols already that likely is good enough for now.

**Reply by maxkatz6** [reply]
> If you are telling me that they don't have WinUI3 + .NET 5/6 

No, I was talking about older APIs with .NET 5/6, such as MapControl from Windows Community Toolkit (using XAML Islands) - that works only on .NET Core 3.1 and older.

It's technically should be possible to use XAML Islands for WinForms or WPF inside of Avalonia app. But again, only on .NET Core 3.1 and older.

**Reply by robloo** [reply]
> No, I was talking about older APIs with .NET 5/6, such as MapControl from Windows Community Toolkit (using XAML Islands) - that works only on .NET Core 3.1 and older.

Ah, sorry, misunderstood. Yea, XAML Islands isn't available at all yet and that is slotted for some unknown date in the future.

**Reply by robloo** [reply]
Mapsui is now MIT license instead of LGPL3 which should make it a lot more usable with Avalonia apps.

**Reply by robloo** [reply]
Mapsui 4.0 (currently in beta) will include Avalonia support:
https://github.com/Mapsui/Mapsui/tree/master/Mapsui.UI.Avalonia

This one can be considered complete as well. Perhaps it should be recommended in the docs some place though.

**Reply by timunie** [reply]
I think awesome avalonia is the right place to mention it

**Comment by maxkatz6** [comment]
Native WebBrowser using NativeControlHost.
Control that will use Cocoa WebView on macOS, IE/Edge WebView on Windows. Not sure if there is any alternative for Linux though (@kekekeks).

**Reply by maxkatz6** [reply]
Right now users can implement their own native WebView on top of NativeControlHost. We even have sample with Cocoa WebView (see [EmbedSample.cs#L77-L91](https://github.com/AvaloniaUI/Avalonia/blob/master/samples/interop/NativeEmbedSample/EmbedSample.cs#L77-L91)).
Another approach is to use Chromium (CEF) based implementation, Avalonia already have couple of them driven by community.

**Reply by Splitwirez** [reply]
Dare I mention the browser monopoly problem...?

**Reply by maxkatz6** [reply]
Nah, it's fine. This monopoly does not really makes it even worse. 
The biggest problem is with Windows, where
1. IE is easy to integrate, but old
2. Edge Legacy was possible to integrate in older .NET Core versions, but is not currently (unless you write all COM interfaces by your own, but I doubt it's any kind of easy approach, unless these interfaces are open sourced and works fine with CsWinRT).
3. Edge Chromium requires specific runtime to be installed on users PC (which is preinstalled only since 2021 updates, in other words - not for that much users yet). Or you can install it with your app (not always possible solution). Or bundle this runtime with your app (why not to use CEF in this case).

**Reply by maxkatz6** [reply]
On macOS problem mostly is with mac API wrappers. I ended up using old MonoMAC, because Xamarin does not work in .NET 5. 
Most likely it won't be a problem with .NET 6 anymore.

**Reply by maxkatz6** [reply]
And if somebody will decide to maintain native WebView, it's unlikely to be maintained by Avalonia team, but more like individual.

**Reply by Splitwirez** [reply]
So...you *don't* think leveraging the renderer from Chromium/Chromium Edge is contributing to the monopoly issue? I'm curious as to why, but...something tells me that's probably a question best saved for Discord DMs...

**Reply by robloo** [reply]
Unlike IE6, chromium is open sourced. So the fact it has become almost a monopoly is not nearly as much of a concern. Linux is a monopoly on high performance scientific computers but nobody minds. This is how it always goes, when a monopoly starts to stagnate new ideas come in which is how IE6 was defeated. The same is true of very large business who become too big to change and recognize good ideas, smaller agile companies eventually defeat them: look at what Avalonia itself is doing because Microsoft is now so big and mismanaged? This cycle is everywhere, even forests. Don't fear an open source monopoly. Eventually it will be replaced when it looses its way. Until then we can all enjoy some standardization. (That said, I use Firefox because it isnt monetizing data).

**Reply by Splitwirez** [reply]
@robloo You've raised a valid point...it's certainly good that Chromium is at least open-source. But it's one thing for there to be a monopoly in a specific niche...it's another thing entirely for there to be a monopoly over something that was once referred to as an "information superhighway" (and still lives up to that description today, arguably far more so than when the term was popularly used). And no, smaller, nimbler companies don't *always* defeat the larger players. Windows has been the desktop OS of choice for most people for a long time now, and that hasn't changed yet.

The internet is a *really* big deal. Sure, a Chromium monopoly wouldn't give Google absolute power over the internet itself...but it *would* give them a great deal of power over the *ability to access* the internet. They could make that codebase do what they see fit (such as continuing to validate the very privacy concerns which you yourself cited as your motivation for maining Firefox), and due to its sheer size, a fork, offshoot, or competitor aiming to allow people to evade those terms would have an ever-growing uphill battle on their hands in their quest do so.

I'm not saying I think you're wrong, but I *am* saying I fear that I might also have a point. As for how this affects Avalonia...well, if Avalonia's default offering for displaying web content is based on Chromium's tech, that could end up meaning that the "default behaviour" for web content in Avalonia (which, let's face it, is likely what most use devs wanting to display web content in Avalonia will use) is to push Google's disregard for user privacy on that front as well.


...the web standards aren't strict enough to make retroactively swapping Chromium's tech for some alternative in a later Avalonia release viable, are they? (e.g. if say, Avalonia 0.11 had a `WebView` that used Chromium tech, but then some kind of crazy Miraclezilla happened to allow Servo to start gaining meaningful traction and Avalonia 0.12/0.13/0.14/etc wanted to change `WebView` to use Servo, that...would likely be a very messy transition for developers who use Avalonia and were already using that `WebView` control following Avalonia 0.11, wouldn't it?)


Again, I apologize for my paranoia...but you know what they say..."you're not paranoid if you're right". I hope I'm wrong about all of this, but I fear my concerns may someday be warranted, at least in part.

**Reply by maxkatz6** [reply]
While it's a good discussion, I don't think it's a good place for it. 
My idea to use native browser was pretty much pragmatic:
1. To not require user to bundle whole CEF, if we can use default one.  
2. Even if user is fine with bundling whole CEF, then Avalonia already has community-driven controls for that.

**Reply by robloo** [reply]
Just putting a link here: https://github.com/maxkatz6/AvaloniaWebView/

**Comment by robloo** [comment]
## FontIcon / PathIcon / BitmapIcon

In addition to the base IconElement per #4087 and #5867

**Reply by pr8x** [reply]
`PathIcon` was added as per https://github.com/AvaloniaUI/Avalonia/pull/4761
